airbyte-bootloader/src/main/java/io/airbyte/bootloader/ProtocolVersionChecker.java

```diff
@@ -0,0 +1,218 @@
+      return Optional.of(targetRange);
+    }
+
+    log.info("Detected an AirbyteProtocolVersion range change from [{}:{}] to [{}:{}]",
+        currentRange.min().serialize(), currentRange.max().serialize(),
+        targetRange.min().serialize(), targetRange.max().serialize());
+
+    final Map<ActorType, Set<UUID>> conflicts = getConflictingActorDefinitions(targetRange);
+
+    if (conflicts.isEmpty()) {
+      log.info("No conflicts");
+      return Optional.of(targetRange);
+    }
+
+    final Set<UUID> destConflicts = conflicts.getOrDefault(ActorType.DESTINATION, new HashSet<>());
+    final Set<UUID> sourceConflicts = conflicts.getOrDefault(ActorType.SOURCE, new HashSet<>());
+
+    if (!supportAutoUpgrade) {
+      // If we do not support auto upgrade, any conflict of used connectors must be resolved before being
+      // able to upgrade the platform.
+      log.warn("The following connectors need to be upgraded before being able to upgrade the platform");
+      formatActorDefinitionForLogging(destConflicts, sourceConflicts).forEach(log::warn);
+      return Optional.empty();
+    }
+
+    final Set<UUID> remainingDestConflicts =
+        projectRemainingConflictsAfterConnectorUpgrades(targetRange, destConflicts, ActorType.DESTINATION);
+    final Set<UUID> remainingSourceConflicts =
+        projectRemainingConflictsAfterConnectorUpgrades(targetRange, sourceConflicts, ActorType.SOURCE);
+
+    if (!remainingDestConflicts.isEmpty() || !remainingSourceConflicts.isEmpty()) {
+      // These set of connectors need a manual intervention because there is no compatible version listed
+      formatActorDefinitionForLogging(remainingDestConflicts, remainingSourceConflicts).forEach(log::warn);
+      return Optional.empty();
+    }
+
+    // These can be auto upgraded
+    destConflicts.removeAll(remainingDestConflicts);
+    sourceConflicts.removeAll(remainingSourceConflicts);
+    log.info("The following connectors will be upgraded");
+    formatActorDefinitionForLogging(destConflicts, sourceConflicts).forEach(log::info);
+    return Optional.of(targetRange);
+  }
+
+  protected Optional<AirbyteVersion> getCurrentAirbyteVersion() throws IOException {
+    return jobPersistence.getVersion().map(AirbyteVersion::new);
+  }
+
+  protected AirbyteProtocolVersionRange getCurrentProtocolVersionRange() throws IOException {
+    Optional<Version> min = jobPersistence.getAirbyteProtocolVersionMin();
+    Optional<Version> max = jobPersistence.getAirbyteProtocolVersionMax();
+
+    if (min.isPresent() != max.isPresent()) {
+      // Flagging this because this would be highly suspicious but not bad enough that we should fail
+      // hard.
+      // If the new config is fine, the system should self-heal.
+      log.warn("Inconsistent AirbyteProtocolVersion found, only one of min/max was found. (min:{}, max:{})",
+          min.map(Version::serialize).orElse(""), max.map(Version::serialize).orElse(""));
```

Comment:
Nit:  consider extracting to a function that can be reused for both `min` and `max`.  Something like:

```
private Function<Optional<Version>, String> versionFunction = v -> v.map(Version::serialize).orElse("");

...

log.warn("... (min:{}, max:{})", versionFunction.apply(min), versionFunction.apply(max));
```

It doesn't reduce the code that much, but does prevent repetition of the logic to extract the version from the optional to avoid potential drift.

Reply:
Yes, this makes sense, but I am thinking of pushing it a bit further and changing the `toString()` implementation and stop using `serialize` everywhere.

The different between the two is about format, `serialize()` gives a `<major>.<minor>.<patch>` while `toString()` returns `Version{version='<major>.<minor>.<patch>', major='<major>', minor='<minor>', patch='<patch>'}`.